source/grid/grid_generator.cc

```diff
@@ -3927,6 +3927,96 @@ namespace GridGenerator
   }
 
 
+  template <int dim, int spacedim>
+  void replicate_triangulation(const Triangulation<dim, spacedim> &basic_tria,
+                               const std::array<Point<spacedim>, 2> &diagonals,
+                               const std::array<unsigned int, dim> &repetitions,
+                               Triangulation<dim, spacedim> &result)
+  {
+    Assert(basic_tria.n_levels()== 1,
+           ExcMessage("The input triangulations must be coarse meshes."));
+    const Point<spacedim> p1=diagonals.front(), p2=diagonals.back();
+
+    // side lengths
+    std::vector<double> sides;
+    for (unsigned int i=0; i<spacedim; ++i)
+      sides.push_back(std::fabs(std::max(p1[i],p2[i])-std::min(p1[i], p2[i])));
+
+    std::vector<unsigned int> reps(repetitions.begin(), repetitions.end());
+    if (reps.size()<3) reps.insert(reps.end(), 3-dim, 1);
+    const unsigned int x_rep=reps[0], y_rep=reps[1], z_rep=reps[2];
+
+    // vertices vector for new triangulation
+    std::vector<Point<spacedim>> vertices(basic_tria.get_vertices());
```

Comment:
You know the final size at this point already. Hence, just use `reserve` as well.